sense-android-library/src/nl/sense_os/datastorageengine/CommonSenseProxy.java

```diff
@@ -0,0 +1,597 @@
+            throw new IOException("getAllDevices: invalid input of sessionID");
+        String params = "&per_page=1000&details=full";
+        return getListForURLAction(URL_DEVICES, params, "devices", sessionID, 1000,"getAllDevices");
+    }
+
+    /**
+     * Add sensor to a device.
+     *
+     * This will create a device if it does not exist yet. This is the only way to create a new device.
+     * A device is uniquely identified by a device ID or by a name and UUID.
+     * This method will use the device ID as stored in the commonsense server.
+     * There is a twin method available that takes a name and UUID instead if the device ID is not available.
+     *
+     * @param csSensorID	CommonSense sensor ID for the sensor. Cannot be empty.
+     * @param deviceType    The name of the device.
+     * @param UUID      	CommonSense device ID for the device. Cannot be empty.
+     * @param sessionID		The sessionID of the current user. Cannot be empty.
+     * @throws IOException, JSONException
+     * @result				Whether or not the sensor was successfully added to the device.
+     */
+    public boolean addSensor(String csSensorID, String deviceType, String UUID, String sessionID)throws IOException, JSONException
+    {
+        if(csSensorID == null || csSensorID.isEmpty() || deviceType == null || deviceType.isEmpty()|| UUID == null || UUID.isEmpty()|| sessionID == null || sessionID.isEmpty())
+            throw new IOException("invalid input of csSensorID or deviceType or UUID or sessionID");
+
+        JSONObject sensor = new JSONObject();
+        sensor.put("type", deviceType);
+        sensor.put("uuid", UUID);
+        JSONObject postData = new JSONObject();
+        postData.put("device", sensor);
+
+        String url = URL_SENSORS + "/" + csSensorID + "/" + URL_SENSOR_DEVICE;
+               url = makeUrlFor(url,null);
+        // perform actual request
+        Map<String, String> response = request(url, postData, sessionID, HTTP_METHOD_POST);
+        int code = checkResponseCode(response.get(RESPONSE_CODE), "addSensor");
+        // check if the response code is 201 CREATED
+        if(code != -2)
+            throw new IOException("Incorrect response of addSensor from CommonSense: " + code);
+        return (code == -2);
+    }
+
+    /**
+     * Upload sensor data to commonsense
+     *
+     * Data can be coming from multiple sensors.
+     *
+     * @param data			JSONArray of datapoints.
+     *                      Each datapoint should be a JSONObject with fields called "sensorID", "value", and "date".
+     *                      These fields will be parsed into the correct form for uploading to commonsense.
+     * @param sessionID		The sessionID of the current user. Cannot be empty.
+     * @throws IOException, JSONException
+     * @result				Whether or not the post of the data was succesfull.
+     */
+    public boolean postData(JSONArray data, String sessionID) throws IOException, JSONException
```

Comment:
@excaliburnan would not throw a JSONException if this has no information for the user